core/src/autogluon/core/utils/feature_selection.py

```diff
@@ -0,0 +1,497 @@
+        candidate_features = self.compute_next_candidate_given_fi(importance_df, prune_threshold, prune_ratio, time_budget_select, weighted)
+
+        # if noise columns exist, they should never be removed
+        if auto_threshold:
+            candidate_features = candidate_features + list(prioritized)
+            importance_df = pd.concat([importance_df, noise_rows])
+
+        feature_selection_time = time.time() - time_start
+        self.time_limit = self.time_limit - feature_selection_time
+        self._fi_time_elapsed = self._fi_time_elapsed + feature_selection_time
+        return candidate_features, importance_df.sort_values(by='importance', axis=0), feature_selection_time
+
+    def compute_next_candidate_given_fi(self, importance_df: pd.DataFrame, prune_threshold: float, prune_ratio: float, time_budget: float,
+                                        weighted: bool) -> List[str]:
+        """
+        Keep features whose importance scores are above threshold or have not yet had a chance to be calculated,
+        as well as some features whose importance scores are below threshold if more than prune_ratio * num features
+        features are below threshold. In the latter case, randomly sample without replacement from features whose
+        importance scores are below threshold until removal candidate configuration that has not yet been tried
+        is encountered. Give higher probability to features whose scores are lower than others when sampling.
+        """
+        time_start = time.time()
+        n_remove = max(1, int(prune_ratio * len(importance_df)))
+        above_threshold_rows = importance_df[(importance_df['importance'] > prune_threshold) | (importance_df['importance'].isna())]
+        below_threshold_rows = importance_df[importance_df['importance'] <= prune_threshold].sort_values(by='importance', axis=0, ascending=True)
+        logger.log(30, f"\tNumber of identified features below prune threshold {round(prune_threshold, 4)}: {len(below_threshold_rows)}/{len(importance_df)}")
+        if len(below_threshold_rows) <= n_remove:
+            acceptance_candidates = above_threshold_rows.index.tolist()
+            self.attempted_removals.add(tuple(below_threshold_rows.index))
+            return acceptance_candidates
+
+        # Try removing features with lowest importance first
+        removal_candidate_rows = below_threshold_rows[:n_remove]
+        removal_candidates = tuple(removal_candidate_rows.index)
+        if removal_candidates not in self.attempted_removals:
+            acceptance_candidates = importance_df[~importance_df.index.isin(removal_candidates)].index.tolist()
+            self.attempted_removals.add(removal_candidates)
+            return acceptance_candidates
+
+        sample_weights = [1/i for i in range(1, len(below_threshold_rows)+1)] if weighted else None
+        while time_budget - (time.time() - time_start) > 0:
+            random_state = self.rng.integers(low=0, high=1e5)
```

Comment:
Do we want this to be non-deterministic?

Reply:
Yeap, because we want to propose a different feature subset to prune every time (in case previous attempts failed to improve the validation score)